src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -2258,6 +2258,362 @@ LogicalResult ONNXConvIntegerOp::inferShapes() {
+  int64_t inputRank = inputTensorTy.getShape().size();
+  if (inputRank != repeatsLength)
+    return emitError("Repeats tensor must have the same length as the input's "
+                     "dimension number.");
+
+  // Check if second argument of TileOp is a constant.
+  auto constantOp = getONNXConstantOp(repeats());
+
+  // Compute output's dimensions: output_dim[i] = input_dim[i] * repeats[i]
+  SmallVector<int64_t, 2> dims(inputRank, -1);
+  if (constantOp) {
+    // 1. Initialize output_dim with values from 'input'.
+    //   output_dim[i] = input[i]
+    for (decltype(inputRank) i = 0; i < inputRank; ++i)
+      dims[i] = inputTensorTy.getShape()[i];
+
+    // 2. Update output_dim using values from 'repeats'.
+    // Do this only for static 'input_dim[i]'.
+    //   if (output_dim[i] != -1) output_dim[i] *= repeats[i]
+    DenseElementsAttr valueAttribute =
+        constantOp.valueAttr().dyn_cast<DenseElementsAttr>();
+    if (!valueAttribute)
+      return emitError("DenseElementsAttr expected");
+    // Get repeat values from valueAttribute.
+    auto valueIt = valueAttribute.getValues<IntegerAttr>().begin();
+    for (int i = 0; i < inputRank; ++i)
+      if (dims[i] != -1)
+        dims[i] *= (*valueIt++).cast<IntegerAttr>().getInt();
+
+    if (valueIt != valueAttribute.getValues<IntegerAttr>().end())
+      return emitError("Constant value must have same length as output's rank");
+  }
+
+  getResult().setType(
+      RankedTensorType::get(dims, inputTensorTy.getElementType()));
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Gather
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXGatherOp::inferShapes() {
+  // Cannot infer shape if no shape exists.
+  if (!data().getType().isa<RankedTensorType>())
+    return emitError("Input tensor not ranked");
+  if (!indices().getType().isa<RankedTensorType>())
+    return emitError("Indices tensor not ranked");
+
+  auto inputShape = data().getType().cast<RankedTensorType>().getShape();
+  auto indicesShape = indices().getType().cast<RankedTensorType>().getShape();
+  int64_t inputRank = inputShape.size();
+  int64_t indicesRank = indicesShape.size();
+
+  if (inputRank < 1)
+    return emitError("Input tensor must have rank >= 1");
+
+  // Read 'axis' attribute.
+  auto axisIndex = axis().getSExtValue();
+  // 'axis' must be in [-rank, rank-1]
+  if (axisIndex < -inputRank || axisIndex >= inputRank)
+    return emitError("Gather axis value out of bound");
+  // Convert a negative axis to a positive axis.
+  if (axisIndex < 0) {
+    axisIndex += inputRank;
+    auto builder = mlir::Builder(getContext());
+    axisAttr(builder.getI64IntegerAttr(axisIndex));
+  }
+
+  // If 'indices' is a constant, check whether its values are valid or not.
+  auto constantOp = getONNXConstantOp(indices());
+  if (constantOp && inputShape[axisIndex] != -1) {
```

Comment:
What will happen for the else of this IF statement?

Reply:
We do nothing. 

Actually, the output's shape does not depend on the actual values of the elements in the `indices` tensor, but the Gather operator's description in ONNX mentions bounds for index values, so I checked them here if it is possible (by using IF statement).

Reply:
I see. That's why I feel weird that there is no else part. Such code is another example for issue 224 I opened yesterday. Your comments are welcome.